ui/StatusQ/src/leftjoinmodel.cpp

```diff
@@ -0,0 +1,219 @@
+        }
+        ++i;
+    }
+
+    m_rightModelRolesOffset = rightRolesOffset;
+    m_roleNames = roleNames;
+
+    connect(m_rightModel, &QAbstractItemModel::dataChanged, this,
+            [this](auto& topLeft, auto& bottomRight, auto& roles) {
+        QVector<int> rolesTranslated;
+
+        if (roles.contains(m_rightModelJoinRole)) {
+            rolesTranslated = m_joinedRoles;
+        } else {
+            rolesTranslated = roles;
+
+            for (auto& role : rolesTranslated)
+                role += m_rightModelRolesOffset;
+        }
+
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), rolesTranslated);
+    });
+
+    disconnect(m_leftModel, &QAbstractItemModel::rowsInserted,
+               this, &LeftJoinModel::initializeIfReady);
+    disconnect(m_rightModel, &QAbstractItemModel::rowsInserted,
+               this, &LeftJoinModel::initializeIfReady);
+
+    connect(m_rightModel, &QAbstractItemModel::rowsRemoved, this, [this] {
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), m_joinedRoles);
+    });
+
+    connect(m_rightModel, &QAbstractItemModel::rowsInserted, this, [this] {
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), m_joinedRoles);
+    });
+
+    connect(this, &QAbstractItemModel::dataChanged, this,
+            [this](auto& topLeft, auto& bottomRight, auto& roles ) {
+        if (roles.contains(m_leftModelJoinRole))
+            emit dataChanged(topLeft, bottomRight, m_joinedRoles);
+    });
+
+    QIdentityProxyModel::setSourceModel(m_leftModel);
+}
+
+QVariant LeftJoinModel::data(const QModelIndex& index, int role) const
+{
+    auto idx = m_leftModel->index(index.row(), 0);
+
+    if (role < m_rightModelRolesOffset)
+        return m_leftModel->data(idx, role);
+
+    QVariant joinRoleLeftValue = m_leftModel->data(idx, m_leftModelJoinRole);
+
+    if (m_lastUsedRightModelIndex.isValid()
+            && m_rightModel->data(m_lastUsedRightModelIndex,
+                                  m_rightModelJoinRole) == joinRoleLeftValue)
+    {
+        return m_rightModel->data(m_lastUsedRightModelIndex,
+                                  role - m_rightModelRolesOffset);
+    }
+
+    int rightModelCount = m_rightModel->rowCount();
+
+    for (int i = 0; i < rightModelCount; i++) {
+        auto rightModelIdx =  m_rightModel->index(i, 0);
+        auto rightJointRoleValue = m_rightModel->data(rightModelIdx,
+                                                      m_rightModelJoinRole);
+
+        if (joinRoleLeftValue == rightJointRoleValue) {
+            m_lastUsedRightModelIndex = rightModelIdx;
+
+            return m_rightModel->data(rightModelIdx,
+                                      role - m_rightModelRolesOffset);
+        }
+    }
+
+    return {};
+}
+
+void LeftJoinModel::setLeftModel(QAbstractItemModel* model)
+{
+    if (m_leftModel == model)
+        return;
+
+    if (m_leftModel != nullptr) {
+        qWarning("Changing left model is not supported!");
```

Comment:
Hmm, how does this work then?

Reply:
Ah I see, so this is a sort of `write-once` property?

Reply:
Yeah, in the future we can consider extending it by changing submodels on the fly. But it requires special care and many additional tests. For now it's big enough :)